autopilot/agent.go

```diff
@@ -448,20 +457,59 @@ func (a *Agent) controller(startingBalance btcutil.Amount) {
 					continue
 				}
 
-				nID := NewNodeID(chanCandidate.PeerKey)
-				pendingOpens[nID] = Channel{
-					Capacity: chanCandidate.ChanAmt,
-					Node:     nID,
-				}
-
 				go func(directive AttachmentDirective) {
-
+					// We'll start out by attempting to
+					// connect to the peer in order to begin
+					// the funding workflow.
 					pub := directive.PeerKey
-					err := a.cfg.ChanController.OpenChannel(
+					err := a.cfg.ConnectToPeer(
+						pub, directive.Addrs,
+					)
+					if err != nil {
+						log.Warnf("Unable to connect "+
+							"to %x: %v",
+							pub.SerializeCompressed(),
+							err)
+						return
+					}
 
-						directive.PeerKey,
-						directive.ChanAmt,
-						directive.Addrs,
+					// If we were succesful, we'll track
+					// this peer in our set of pending
+					// opens. We do this here to ensure we
+					// don't stall on selecting new peers if
+					// the connection attempt happens to
+					// take too long.
+					pendingMtx.Lock()
+					if uint16(len(pendingOpens))+1 >
+						a.cfg.MaxPendingOpens {
+
+						// Since we've reached our max
+						// number of pending opens,
+						// we'll disconnect this peer
+						// and exit.
+						err = a.cfg.DisconnectPeer(
+							pub,
+						)
+						if err != nil {
+							log.Warnf("Unable to "+
+								"disconnect peer "+
+								"%x: %v",
+								pub.SerializeCompressed(),
+								err)
+						}
```

Comment:
Mutex is never unlocked. Otherwise would halt all autopilot progress.

Reply:
Nice catch, fixed!